### ISSUE TITLE
Incorrect gradients on constant function

### ISSUE THREAD

**Issue by sracaniere** [google/jax#15588]
### Description

I have a smooth function that is exactly 0 outside of a certain interval, but jax can give me non-zero gradients on those values. Here is the code I used to get this:
```python
import collections
from typing import Any, Callable, Optional, Sequence, Union

import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np

jax.config.update('jax_enable_x64', True)

def _bump_right(x: jax.Array) -> jax.Array:
  assert x.dtype == jnp.float64
  eps = 1e-2
  # `xx` is used in `-1/xx`. We prevent it from crossing 0 to avoid
  # infinities and NaN in the backward pass when computing gradients.
  xx = jnp.maximum(x, eps)
  return jnp.where(x < 0.0, 0.0, jnp.exp(-1.0 / xx))


def _smooth_transition_right(x: jax.Array) -> jax.Array:
  """A smooth increasing function, exactly 0 on x<0 and 1 on x>1."""
  bx = _bump_right(x)
  bout = bx / (bx + _bump_right(1.0 - x))
  return bout


def well(
    x: jax.Array,
    inner_radius: chex.Numeric,
    outer_radius: chex.Numeric,
) -> jax.Array:
  """A smooth even function that is 0 on [-inner_radius, inner_radius] and 1
  outside of [-outer_radius, outer_radius]. See plot below."""
  size_transition = outer_radius - inner_radius
  return _smooth_transition_right(
      (x - inner_radius) / size_transition
  ) + _smooth_transition_right((-x - inner_radius) / size_transition)


r = jnp.arange(-2., 2., 0.01)
# Run in a cola to see these plots
plt.plot(r, well(r, 0.7, 1.05))
plt.show()
plt.plot(r, jnp.equal(well(r, 0.7, 1.05), 0.))
plt.plot(r, jnp.equal(well(r, 0.7, 1.05), 1.))
plt.show()

# Compute derivatives
def forward_this_cell(r):
  """Defines the selected PINN model."""
  expansion = jnp.sinh(r) ** 4

  # The `expansion_weight` are exactly 0 when abs(r) > 1. In particular, the
  # output of this function is constant equal to 0 when abs(r) > 1.
  expansion_weight = 1.0 - well(
      r,
      inner_radius=0.7,
      outer_radius=1.,
  )
  # Uncomment this line, and the problem disappears
  # expansion_weight = jnp.where(
  #     jnp.equal(expansion_weight, 0.0), 0.0, expansion_weight
  # )
  return expansion * expansion_weight


def get_der(z):
  """1st order derivative."""
  def f(zz):
    return jnp.sum(forward_this_cell(zz))
  return jax.grad(f)(z)


rng_seq = hk.PRNGSequence(2609785)
z_data = jnp.array(
    [[0.1], [8.72325556], [22.03669951]]
)
# In the code below, we expect all derivatives wrt z_data[1] and z_data[2] to
# be exactly 0. Instead, we see some very large numbers.
print('z_data\n', z_data, '\n')
print('ref der\n', get_der(z_data), '\n')
for _ in range(16):
  noise = 1e-9 * jax.random.normal(next(rng_seq), z_data.shape, z_data.dtype)
  print('der')
  print(get_der(z_data + noise))
  print('')
```

### Which accelerator(s) are you using?

CPU/GPU

**Comment by jakevdp** [comment]
Thanks for the report – I'm having trouble understanding what the relevant pieces of this are. The output (after adding the missing `haiku` and `chex` imports) is two plots along with the following text:

<details>
<summary>output</summary>

```
z_data
 [[ 0.1       ]
 [ 8.72325556]
 [22.03669951]] 

ref der
 [[ 4.04016036e-03]
 [ 0.00000000e+00]
 [-1.53422891e+18]] 

der
[[4.04016042e-03]
 [7.01457223e-05]
 [0.00000000e+00]]

der
[[ 4.04016042e-03]
 [ 0.00000000e+00]
 [-1.53422891e+18]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[4.04016033e-03]
 [7.01457223e-05]
 [0.00000000e+00]]

der
[[ 4.04016039e-03]
 [-7.01457223e-05]
 [ 1.53422891e+18]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[4.04016034e-03]
 [7.01457223e-05]
 [0.00000000e+00]]

der
[[ 4.04016009e-03]
 [-7.01457223e-05]
 [ 0.00000000e+00]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[ 4.04016057e-03]
 [-7.01457223e-05]
 [-1.53422891e+18]]

der
[[4.04016041e-03]
 [7.01457223e-05]
 [0.00000000e+00]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[0.00404016]
 [0.        ]
 [0.        ]]

der
[[4.04016038e-03]
 [0.00000000e+00]
 [1.53422891e+18]]
```

</details>

It's not clear to me which output I'm supposed to be looking at, and what values in this output you expect to be different.

Perhaps you could spend a few minutes removing pieces of the code that are not immediately relevant to the bug? In other words, make a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example). Thanks!

**Comment by sracaniere** [comment]
Thanks for looking into this, and sorry about the missing imports.

This is already a minimal example. There are two comments in the code that explain the problem. The comment
```python
# In the code below, we expect all derivatives wrt z_data[1] and z_data[2] to
# be exactly 0. Instead, we see some very large numbers.
```
means that the expected output of every iteration of the for loop should look like
```
der
[[0.00404016]
 [0.        ]
 [0.        ]]
```
Note the two 0.

From the output you generated, I can see that you reproduced the issue since you have non zero values.

The second relevant comment in my code is inside the `forward_this_cell` funtion:
```
  # Uncomment this line, and the problem disappears
  # expansion_weight = jnp.where(
  #     jnp.equal(expansion_weight, 0.0), 0.0, expansion_weight
  # )
```
If you uncomment these 3 lines and rerun the code, you will see that indeed all outputs are as expected with two 0 as above.

**Comment by jakevdp** [comment]
OK, thanks. So we could remove the loop and just print a single value to demonstrate the bug, yes? IMHO that would be much clearer.

**Comment by sracaniere** [comment]
Yes, you could do that, but then you'd loose the information that tiny variations (or the order of 1e-9) to the inputs leads to sometimes huge numbers (e.g. with exponent of 35) instead of 0, sometimes to small errors, and sometimes to 0 indeed.

**Comment by jakevdp** [comment]
Sure, but the relevant info is not that it sometimes produces the correct results, the relevant info is the inputs where it produces incorrect results. As it is, anyone trying to fix this bug will have to do the work of removing irrelevant pieces and reducing your long repro to a short one that can be debugged – it's helpful to package maintainers if you do that for them.

Just trying to help you to raise the most easily actionable issues – if you don't have the time to do it on this one, then we'll take care of it, but please just keep that in mind for the future!

**Comment by sracaniere** [comment]
I do have time for this. I actually spent quite a lot of time isolating that bug from a much larger piece of code.

Here is the cleaned up version:
```python
import collections
from typing import Any, Callable, Optional, Sequence, Union

import chex
import haiku as hk
import jax
import jax.numpy as jnp
import matplotlib.pyplot as plt
import numpy as np

jax.config.update('jax_enable_x64', True)

def _bump_right(x: jax.Array) -> jax.Array:
  assert x.dtype == jnp.float64
  eps = 1e-2
  # `xx` is used in `-1/xx`. We prevent it from crossing 0 to avoid
  # infinities and NaN in the backward pass when computing gradients.
  xx = jnp.maximum(x, eps)
  return jnp.where(x < 0.0, 0.0, jnp.exp(-1.0 / xx))


def _smooth_transition_right(x: jax.Array) -> jax.Array:
  """A smooth increasing function, exactly 0 on x<0 and 1 on x>1."""
  bx = _bump_right(x)
  bout = bx / (bx + _bump_right(1.0 - x))
  return bout


def well(
    x: jax.Array,
    inner_radius: chex.Numeric,
    outer_radius: chex.Numeric,
) -> jax.Array:
  """A smooth even function that is 0 on [-inner_radius, inner_radius] and 1
  outside of [-outer_radius, outer_radius]. See plot below."""
  size_transition = outer_radius - inner_radius
  return _smooth_transition_right(
      (x - inner_radius) / size_transition
  ) + _smooth_transition_right((-x - inner_radius) / size_transition)


r = jnp.arange(-2., 2., 0.01)
# Run in a cola to see these plots
plt.plot(r, well(r, 0.7, 1.05))
plt.show()
plt.plot(r, jnp.equal(well(r, 0.7, 1.05), 0.))
plt.plot(r, jnp.equal(well(r, 0.7, 1.05), 1.))
plt.show()

# Compute derivatives
def forward_this_cell(r):
  """Defines the selected PINN model."""
  expansion = jnp.sinh(r) ** 4

  # The `expansion_weight` are exactly 0 when abs(r) > 1. In particular, the
  # output of this function is constant equal to 0 when abs(r) > 1.
  expansion_weight = 1.0 - well(
      r,
      inner_radius=0.7,
      outer_radius=1.,
  )
  # Uncomment this line to fix the incorrect derivative
  # expansion_weight = jnp.where(
  #     jnp.equal(expansion_weight, 0.0), 0.0, expansion_weight
  # )
  return expansion * expansion_weight


def get_der(z):
  """1st order derivative."""
  def f(zz):
    return jnp.sum(forward_this_cell(zz))
  return jax.grad(f)(z)


rng_seq = hk.PRNGSequence(2609785)
z_data = jnp.array([[8.72325556], [22.03669951]])
# The correct value printed below should be [[0.], [0.]], but I got.
# [[ 0.00000000e+00]
#  [-1.53422891e+18]]
# Check `forward_this_cell` above to find some lines to uncomment in
# order to workaround the problem. This is only a workaround though.
print(get_der(z_data))
```

**Comment by jakevdp** [comment]
Thanks - next time I'd suggest removing the plotting code: it's not relevant to the reported bug.

**Comment by jakevdp** [comment]
I spent some time minimizing your code and got a more actionable reproduction:
```python
import jax
import jax.numpy as jnp

jax.config.update('jax_enable_x64', True)

def f(r):
  b1 = jnp.where(r - 0.7 < 1e-2, 0.0, jnp.exp(-1.0 / (r - 0.7)))
  b2 = jnp.where(1.0 - r < 1e-2, 0.0, jnp.exp(-1.0 / (1.0 - r)))
  return jnp.sum(jnp.sinh(r) ** 4 * (1.0 - b1 / (b1 + b2)))

# function outputs exactly zero for all input values beyond 1:
y = jax.vmap(f)(jnp.linspace(1, 30, 1000))
print(jnp.all(y == 0))
# True

# But the gradient shows strange local variation:
print(jax.grad(f)(22.02))
# 0.0

print(jax.grad(f)(22.03))
# 4.952075370762036e+18
```

**Comment by jakevdp** [comment]
I think the issue here is probably coming from
```python
bout = bx / (bx + _bump_right(1.0 - x))
```
which for `x > 0` is effectively:
```python
bx = jnp.exp(-1.0 / x)
bout = bx / (bx + 0.0)
```
When `x` is large, you're dividing two very small numbers, which can lead to issues when working with floating point arithmetic.

Given this, I don't think this is exposing any issue in JAX itself, but rather suggests that you need to code your function with floating point roundoff error in mind. What do you think?

**Comment by jakevdp** [comment]
Something about the `1 - exp(-1/x) / exp(-1/x)` term is causing floating point cancellation issues within the gradient computation - I'm still not sure exactly what the problem could be.

**Comment by sracaniere** [comment]
> I suspect this is a matter of treating floating point arithmetic
That is only a possible explanation in the code you extracted, but not in the code I provided. Your code removes a crucial multiplication `expansion * expansion_weight`. In that multiplication, we have that `expansion_weight` is exactly 0 (not just close to 0). The gradient should therefore be exactly 0, as long as the gradients coming from `expansion` do not contain infinities. If there were numerical instabilities in the computation of `expansion` and some infinities occurred during the gradient computation, then a NaN in the output would make sense. But that's not what we are seeing.

Finally, I had left some code in a comment in my code offering a workaround:
```python
expansion_weight = jnp.where(
    jnp.equal(expansion_weight, 0.0), 0.0, expansion_weight
)
```
This code replaces `expansion_weight` with 0 when `expansion_weight` is already 0. If the problem indeed came from numerical instabilities you see in your code, then my rather strange workaround should not work.

**Comment by sracaniere** [comment]
Thank.

Interestingly, you don't even need the term `exp(-1 / r)` or any `sinh`, you can do:
```python
import jax
import jax.numpy as jnp

jax.config.update('jax_enable_x64', True)


def f(r):
  zero = 1 - r / r
  return 1.65163625e+38 * zero

# function evaluates to zero for all positive r:
y = f(jnp.linspace(1, 30, 1000))
print(jnp.all(y == 0))
# True

# But the gradient is not 0 and shows strange local variation:
print(jax.grad(f)(22.0))
# 0.0

print(jax.grad(f)(22.1))
# -1.1805916207174113e+21
```

So, the problem is just the gradient of `1 - r/r`. And indeed, if we remove the multiplication by the large number above, we see small non-zero gradients:
```python
import jax
import jax.numpy as jnp

jax.config.update('jax_enable_x64', True)


def f(r):
  return 1 - r / r

# function evaluates to zero for all positive r:
y = f(jnp.linspace(1, 30, 1000))
print(jnp.all(y == 0))
# True

# But the gradient is not 0 and shows strange local variation:
print(jax.grad(f)(22.0))
# 0.0

print(jax.grad(f)(22.1))
# -6.938893903907228e-18
```

This looks like just a small 1e-18 error, but when we multiply by a large enough number, we get a large error.

So the question now is: is it a numerical issue that the constant function `1 - r / r` returns non-zero gradients? Here, I don't know enough about the inner working of auto-diff to know if this kind of problems can be avoided.

**Comment by jakevdp** [comment]
Ah, that's really interesting! Good find.

So this is the issue:
```python
import jax
import jax.numpy as jnp

jax.config.update('jax_enable_x64', True)

def f(x):
  return 1 - x / x

print(jax.grad(f)(22.04))
# -6.938893903907228e-18
```
Your question is "why isn't the gradient zero?". My answer is that the gradient **is** zero, it's just that in floating point, due to roundoff error, sometimes zero is spelled `-6.938893903907228e-18`. Here's a classic example:
```python
print(0.3 - 0.2 - 0.1)
# -2.7755575615628914e-17
```
You can read more about this issue is here: [Is floating point math broken?](https://stackoverflow.com/questions/588004/is-floating-point-math-broken)

As a user of floating-point math, it's up to you to make sure you're not multiplying this kind of inexact zero by a large number. 

If you're curious about the mechanics of why this happens in `jax.grad`, we can explore the [jaxpr](https://jax.readthedocs.io/en/latest/jaxpr.html) of the gradient function:
```python
print(jax.make_jaxpr(jax.grad(f))(x))
```
```
{ lambda ; a:f64[]. let
    b:f64[] = div a a
    c:f64[] = integer_pow[y=-2] a
    _:f64[] = sub 1.0 b
    d:f64[] = neg 1.0
    e:f64[] = mul d c
    f:f64[] = mul e a
    g:f64[] = neg f
    h:f64[] = div d a
    i:f64[] = add_any h g
  in (i,) }
```
Looking at this, it basically encodes this function:
```python
def gradf(x):
  return (1 / x) - (x / x ** 2)
```
which makes sense if you think about how you would write out the derivative of `1 - x / x`. And evaluating this function gives you a result that is consistent with zero, to within standard floating point precision:
```python
print(gradf(22.04))
# 6.938893903907228e-18
```
So what can you do? I think you have two options here:

1. re-express your function so that it doesn't rely on exact cancellation that does not occur in floating point arithmetic
2. isolate the part of your function that is causing the problem, and write a [custom_jvp](https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html) to evaluate the derivative in a way that doesn't produce problematic errors around zero.
